src/db/connection.rs

```diff
@@ -0,0 +1,106 @@
+use crate::db::sql_types::*;
+use diesel::pg::PgConnection;
+use rocket::fairing::Fairing;
+use rocket_contrib::databases::r2d2::{Pool, PooledConnection};
+use rocket_contrib::databases::Poolable;
+
+type Connection = PooledConnection<<PgConnection as Poolable>::Manager>;
+
+/// The request guard type.
+pub struct RequestContext {
+    pub conn: Connection,
+}
+
+/// The pool type.
+pub struct ConnectionPool(Pool<<PgConnection as Poolable>::Manager>);
+
+impl From<Connection> for RequestContext {
+    fn from(conn: Connection) -> Self {
+        Self { conn }
+    }
+}
+impl RequestContext {
+    /// Get a DatabaseContext from this struct that can be used to connect
+    /// to the db
+    pub fn database_context(
+        &self,
+        action: ActionType,
+        modifier: ActionModifier,
+    ) -> DatabaseContext {
+        return DatabaseContext::from(&self.conn, action, modifier);
+    }
+
+    /// Returns a fairing that initializes the associated database
+    /// connection pool.
+    pub fn fairing() -> impl Fairing {
+        ::rocket::fairing::AdHoc::on_attach("\'postgres\' Database Pool", |rocket| {
+            let pool = ::rocket_contrib::databases::database_config("postgres", rocket.config())
+                .map(PgConnection::pool);
```

Comment:
I'm moderately sure you can use a `flat_map` here or `and_then` or something to get rid of the nested result.

Reply:
I'm not sure what to do about this. The most recent commit uses two separate match expressions, and prints error messages using Rocket's logger.

Reply:
I'd like to do use the `?` operator, but that doesn't work because the function needs to return a `Result<Rocket, Rocket>`, so I can't just propogate errors up. Additionally, because there are 2 separate error types we need to handle, I think the original writers of the code wanted to preserve error data (I got the code from `cargo expand db::connection`, and the proc_macro that generated it can be found [here](https://github.com/SergioBenitez/Rocket/blob/master/contrib/codegen/src/database.rs))